packages/front-end/components/Settings/EditDataSource/ExperimentAssignmentQueries/AddEditExperimentAssignmentQueryModal.tsx

```diff
@@ -97,6 +113,123 @@ export const AddEditExperimentAssignmentQueryModal: FC<EditExperimentAssignmentQ
     return null;
   }
 
+  const handleTestQuery = async () => {
+    setTestQueryResults(null);
+
+    try {
+      const requiredColumns = getRequiredColumns(
+        userEnteredUserIdType,
+        userEnteredDimensions,
+        userEnteredHasNameCol
+      );
+      validateSQL(userEnteredQuery, requiredColumns);
+
+      const res: TestQueryResults = await apiCall("/query/test", {
+        method: "POST",
+        body: JSON.stringify({
+          query: userEnteredQuery,
+          datasourceId: dataSource.id,
+        }),
+      });
+
+      if (res.error) {
+        setTestQueryResults({ error: res.error });
+        return;
+      }
+
+      const warningsArr = [];
+      const optionalColumns = [];
+      const returnedColumns = [];
+      const namedCols = ["experiment_name", "variation_name"];
+      const userIdTypes = dataSource.settings.userIdTypes?.map(
+        (type) => type.userIdType || []
+      );
+
+      if (res.results.length > 0) {
+        for (const column in res.results[0]) {
+          if (
+            !requiredColumns.includes(column) &&
+            !namedCols.includes(column) &&
+            !userIdTypes.includes(column)
+          ) {
+            optionalColumns.push(column);
+          }
+
+          returnedColumns.push(column);
+        }
+
+        // If the user didn't check the box for includesNameColumns, but included
```

Comment:
If you have suggestions on a different way to handle the logic from line 161 to 191, I'm open to it. It works, but it's not the easiest logic to follow.